pypulseq/make_extended_trapezoid_area.py

```diff
@@ -35,100 +33,93 @@ def make_extended_trapezoid_area(
     grad : SimpleNamespace
         Extended trapezoid event.
     times : numpy.ndarray
+        Time points of the extended trapezoid.
     amplitude : numpy.ndarray
+        Amplitude values of the extended trapezoid.
     """
-    SR = system.max_slew * 0.99
-
-    Tp = 0
-    obj1 = (
-        lambda x: (
-            area - __testGA(x, 0, SR, system.grad_raster_time, grad_start, grad_end)
+    slew_rate = system.max_slew * 0.99
+    max_grad = system.max_grad * 0.99
+    raster_time = system.grad_raster_time
+
+    def _to_raster(time: float) -> float:
+        return np.ceil(time / raster_time) * raster_time
+
+    def _calc_area_and_times(grad_amp, flat_time, slew_rate, grad_start, grad_end):
+        time_ramp_up = _calc_ramp_up_time(grad_amp, slew_rate, grad_start)
+        time_ramp_down = _calc_ramp_down_time(grad_amp, slew_rate, grad_end)
+        area = _calc_area(grad_amp, time_ramp_up, flat_time, time_ramp_down, grad_start, grad_end)
+        return area
+
+    def _calc_area(grad_amp, time_ramp_up, flat_time, time_ramp_down, grad_start, grad_end):
+        return (
+            0.5 * time_ramp_up * (grad_amp + grad_start)
+            + grad_amp * flat_time
+            + 0.5 * (grad_amp + grad_end) * time_ramp_down
         )
-        ** 2
-    )
-    arr_res = [
-        minimize(fun=obj1, x0=-system.max_grad, method="Nelder-Mead"),
-        minimize(fun=obj1, x0=0, method="Nelder-Mead"),
-        minimize(fun=obj1, x0=system.max_grad, method="Nelder-Mead"),
-    ]
-    arr_res = np.array([(*res.x, res.fun) for res in arr_res])
-    Gp, obj1val = arr_res[:, 0], arr_res[:, 1]
-    i_min = np.argmin(obj1val)
-    Gp = Gp[i_min]
-    obj1val = obj1val[i_min]
-
-    if obj1val > 1e-3 or abs(Gp) > system.max_grad:  # Search did not converge
-        Gp = math.copysign(system.max_grad, Gp)
-        obj2 = (
-            lambda x: (
-                area
-                - __testGA(Gp, x, SR, system.grad_raster_time, grad_start, grad_end)
-            )
-            ** 2
-        )
-        res2 = minimize(fun=obj2, x0=0, method="Nelder-Mead")
-        T, obj2val = *res2.x, res2.fun
-        assert obj2val < 1e-2
 
-        Tp = math.ceil(T / system.grad_raster_time) * system.grad_raster_time
-
-        # Fix the ramps
-        Tru = (
-            math.ceil(abs(Gp - grad_start) / SR / system.grad_raster_time)
-            * system.grad_raster_time
-        )
-        Trd = (
-            math.ceil(abs(Gp - grad_end) / SR / system.grad_raster_time)
-            * system.grad_raster_time
-        )
-        obj3 = lambda x: (area - __testGA1(x, Tru, Tp, Trd, grad_start, grad_end)) ** 2
+    def _calc_ramp_up_time(grad_amp, slew_rate, grad_start):
```

Comment:
`_calc_ramp_up_time` and `_calc_ramp_down_time` are the same function, can just define one `_calc_ramp_time`.